packages/indexer/src/sinks/drizzle/transaction.ts

```diff
@@ -0,0 +1,106 @@
+import type { Cursor } from "@apibara/protocol";
+import {
+  type ExtractTablesWithRelations,
+  type SQL,
+  type Subquery,
+  type TablesRelationalConfig,
+  sql,
+} from "drizzle-orm";
+import type {
+  PgDeleteBase,
+  PgInsertBuilder,
+  PgInsertValue,
+  PgQueryResultHKT,
+  PgSelectBuilder,
+  PgTable,
+  PgTransaction,
+  PgUpdateBuilder,
+  PgUpdateSetSource,
+  SelectedFields,
+} from "drizzle-orm/pg-core";
+import type { PgViewBase } from "drizzle-orm/pg-core/view-base";
+import type { Int8Range } from "./Int8Range";
+import { getDrizzleCursor } from "./utils";
+
+export class DrizzleSinkTransaction<
+  TQueryResult extends PgQueryResultHKT,
+  TFullSchema extends Record<string, unknown> = Record<string, never>,
+  TSchema extends
+    TablesRelationalConfig = ExtractTablesWithRelations<TFullSchema>,
+> {
+  constructor(
+    private db: PgTransaction<TQueryResult, TFullSchema, TSchema>,
+    private endCursor?: Cursor,
+  ) {}
+
+  insert<TTable extends PgTable>(
+    table: TTable,
+  ): PgInsertBuilder<TTable, TQueryResult> {
+    const originalInsert = this.db.insert(table);
+    // TODO
+    return {
+      ...originalInsert,
+      values: (values: PgInsertValue<TTable> | PgInsertValue<TTable>[]) => {
+        const cursoredValues = (Array.isArray(values) ? values : [values]).map(
+          (v) => ({
+            ...v,
+            _cursor: getDrizzleCursor(this.endCursor) as Int8Range,
+          }),
+        );
+        // biome-ignore lint/suspicious/noExplicitAny: <explanation>
+        return originalInsert.values(cursoredValues as any);
+      },
+    } as PgInsertBuilder<TTable, TQueryResult>;
+  }
+
+  update<TTable extends PgTable>(
```

Comment:
I think `update` and `delete` will benefit from creating a subclass of the original drizzle class and then propagating the cursor by adding an extra property.

Something like

```ts
export class SinkPgUpdateBuilder<TTable extends PgTable, TQueryResult extends PgQueryResultHKT> extends PgUpdateBuilder<TTable, TQueryResult> {
 	constructor(
		private table: TTable,
		private session: PgSession,
		private dialect: PgDialect,
		private withList?: Subquery[],
   private cursor: Int8Range, // or whatever type it is
	) {}

  // Notice the return type is `Sink...`
	set(values: PgUpdateSetSource<TTable>): SinkPgUpdateBase<TTable, TQueryResult> {
		return new SinkPgUpdateBase<TTable, TQueryResult>(
			this.table,
			mapUpdateSet(this.table, values),
			this.session,
			this.dialect,
			this.withList,
		);
	}
}
```

Then `db` returns it like

```ts
return new SinkPgUpdateBuilder(table, this.db.session, this.db.dialect);
```

Reply:
Not sure it's possible but I think it will be easier to maintain